module/input/SensorFilter/src/SensorFilter.cpp

```diff
@@ -165,33 +160,133 @@ namespace module::input {
+                    Eigen::Vector3d gyro  = Eigen::Vector3d::Zero();
+                    Eigen::Vector3d rMFt  = Eigen::Vector3d::Zero();
+                    auto filtered_sensors = std::make_unique<Sensors>();
+
+                    for (const auto& s : sensors) {
+                        // Accumulate accelerometer and gyroscope readings
+                        acc += Eigen::Vector3d(s->accelerometer.x, s->accelerometer.y, s->accelerometer.z);
+                        gyro += Eigen::Vector3d(s->gyroscope.x, s->gyroscope.y, s->gyroscope.z);
+
+                        // Make sure we have servo positions
+                        for (uint32_t i = 0; i < 20; ++i) {
+                            auto& original = utility::platform::getRawServo(i, *s);
+                            // Add the sensor values to the system properly
+                            filtered_sensors->servo.push_back({0,
+                                                               i,
+                                                               original.torque_enabled,
+                                                               original.p_gain,
+                                                               original.i_gain,
+                                                               original.d_gain,
+                                                               original.goal_position,
+                                                               original.moving_speed,
+                                                               original.present_position,
+                                                               original.present_speed,
+                                                               original.load,
+                                                               original.voltage,
+                                                               float(original.temperature)});
+                        }
+
+                        // Calculate forward kinematics
+                        const auto Htx = calculateAllPositions(model, *filtered_sensors);
+                        for (const auto& entry : Htx) {
+                            filtered_sensors->Htx[entry.first] = entry.second.matrix();
+                        }
+
+                        // Calculate the average length of both legs from the torso and accumulate this measurement
+                        const Eigen::Affine3d Htr(filtered_sensors->Htx[ServoID::R_ANKLE_ROLL]);
+                        const Eigen::Affine3d Htl(filtered_sensors->Htx[ServoID::L_ANKLE_ROLL]);
+                        const Eigen::Vector3d rTFt = (Htr.translation() + Htl.translation()) * 0.5;
+
+                        // Accumulator CoM readings
+                        rMFt += calculateCentreOfMass(model, filtered_sensors->Htx).head<3>() + rTFt;
+                    }
+
+                    // Average all accumulated readings
+                    acc /= double(sensors.size());
```

Comment:
This assumes sensors.size() is greater than zero. Can that ever possibly happen? can this code run ever if no sensors messages are available?

Reply:
I dont think `Last<n>` will trigger with no emits?

Reply:
[docs suggest otherwise??](https://nuclear.readthedocs.io/en/latest/dsl.html#last)
@TrentHouliston @CMurtagh-LGTM please confirm or deny

Reply:
Can you trigger if emit doesn't exist..? docs suggest `last [0-n]` but by definition if you 'trigger' This will enact the execution of a task whenever T is emitted into the system.

Reply:
You will always have one lement in a `Last<...>` since the first time it's triggered will be an element and it won't be triggered before that.